packages/common/src/audius-query/createRequestBatcher.ts

```diff
@@ -0,0 +1,140 @@
+import { omit, keyBy } from 'lodash'
+import objectHash from 'object-hash'
+
+import { ID } from '~/models'
+import { Nullable } from '~/utils/typeUtils'
+
+import { AudiusQueryContextType } from './AudiusQueryContext'
+import { schemas } from './schema'
+import { EndpointConfig } from './types'
+
+// Requests occuring in this period will be batched
+const BATCH_PERIOD_MS = 10
+
+type RequestQueueItem<Args, Data> = {
+  fetchArgs: Args
+  context: AudiusQueryContextType
+  resolve: (value: Data) => void
+  reject: (error: unknown) => void
+}
+
+type RequestGroup = {
+  queue: RequestQueueItem<any, any>[]
+  timer: Nullable<NodeJS.Timeout>
+}
+
+/**
+ * Create a request batcher that batches requests for the same endpoint
+ */
+export const createRequestBatcher = () => {
+  const requestGroups: Record<string, RequestGroup> = {}
+
+  const fetch =
+    <Args, Data>(
+      endpointName: string,
+      endpointConfig: EndpointConfig<Args, Data>
+    ) =>
+    (fetchArgs: Args, context: AudiusQueryContextType) => {
+      return new Promise<Data>((resolve, reject) => {
+        const idArgKey = endpointConfig.options.idArgKey ?? 'id'
+
+        // Hash the request args and endpointName to determine the request group
+        const requestGroupKey = objectHash({
+          args: omit(fetchArgs ?? {}, idArgKey),
+          endpointName
+        })
+
+        requestGroups[requestGroupKey] = requestGroups[requestGroupKey] ?? {
+          queue: [],
+          timer: null
+        }
+
+        const requestGroup = requestGroups[requestGroupKey]
+
+        // Queue the request
+        requestGroup.queue = [
+          ...requestGroup.queue,
+          { fetchArgs, context, resolve, reject }
+        ]
+
+        // Set up a timer to perform the batch request
+        if (!requestGroup.timer) {
+          requestGroup.timer = setTimeout(
+            () => performBatch(requestGroup, endpointName, endpointConfig),
+            BATCH_PERIOD_MS
+          )
+        }
+      })
+    }
+
+  const performBatch = async <Args, Data>(
+    requestGroup: RequestGroup,
+    endpointName: string,
+    endpointConfig: EndpointConfig<Args, Data>
+  ) => {
+    const idArgKey = endpointConfig.options.idArgKey ?? 'id'
+    const queuedRequests = [...requestGroup.queue] as RequestQueueItem<
```

Comment:
ooh, what if the `batchFetch` has a max entity count?
If `getTracks` will only give us say 50 tracks, we'll end up with no results for any track over the endpoint's own batch limit. The promises for any tracks beyond 50 will be rejected.

Reply:
We can handle this easily here by only slicing the first `N` items into queuedRequests and leaving the rest with a timer set to 0, or by recursively calling `performBatch`

Reply:
Good thinking, our endpoints have limit 100 by default. I figured we probably won't hit that limit any time soon but I can go ahead and add some logic to handle this case